go/ct/gen/interval_solver.go

```diff
@@ -0,0 +1,177 @@
+package gen
+
+import (
+	"fmt"
+	"strings"
+
+	"golang.org/x/exp/constraints"
+	"pgregory.net/rand"
+)
+
+type IntervalSolver[T constraints.Integer] struct {
+	intervals []interval[T]
+}
+
+func NewIntervalSolver[T constraints.Integer](min, max T) *IntervalSolver[T] {
+	return &IntervalSolver[T]{
+		intervals: []interval[T]{{min, max}},
+	}
+}
+
+func (s *IntervalSolver[T]) Exclude(min, max T) {
+	if max < min {
+		return
+	}
+	res := []interval[T]{}
+	for _, current := range s.intervals {
+		newLeft := interval[T]{current.low, min - 1}
+		newRight := interval[T]{max + 1, current.high}
+
+		switch current.getRelationTo(&interval[T]{min, max}) {
+		case isEqual:
+			continue // we remove this interval
+		case isSubset:
+			res = append(res, newLeft, newRight)
+		case isSuperset:
+			continue // the current interval is dropped
+		case intersectsWithStart:
+			res = append(res, newRight)
+		case intersectsWithEnd:
+			res = append(res, newLeft)
+		default:
+			res = append(res, current)
+		}
+
+	}
+	s.intervals = res
+}
+
+func (s *IntervalSolver[T]) Contains(value T) bool {
+	for _, interval := range s.intervals {
+		if interval.contains(value) {
+			return true
+		}
+	}
+	return false
+}
+
+func (s *IntervalSolver[T]) AddEqualityConstraint(value T) {
+	s.AddLowerBoundary(value)
+	s.AddUpperBoundary(value)
+}
+
+func (s *IntervalSolver[T]) AddLowerBoundary(value T) {
+	if len(s.intervals) == 0 {
+		return
+	}
+	min := s.intervals[0].low
+	s.Exclude(min, value-1)
+}
+
+func (s *IntervalSolver[T]) AddUpperBoundary(value T) {
+	if len(s.intervals) == 0 {
+		return
+	}
+	max := s.intervals[len(s.intervals)-1].high
+	s.Exclude(value+1, max)
+}
+
+func (s *IntervalSolver[T]) IsSatisfiable() bool {
+	return len(s.intervals) > 0
+}
+
+func (s *IntervalSolver[T]) Generate(rnd *rand.Rand) (T, error) {
+	if len(s.intervals) == 0 {
+		return 0, ErrUnsatisfiable
+	}
+
+	domainSize := uint64(0)
+	for _, interval := range s.intervals {
+		domainSize += uint64(interval.high - interval.low + 1)
+	}
+
+	if domainSize == 0 { // the domain is the full uint64 domain
```

Comment:
If domain size is less than 0, because of invalid/empty intervals, there should be an error

Reply:
because `domainSize` is a `uint64` it can not be negative, it would be an overflow, which could easily be confused with a big interval, for this a check was added to verify that the interval is not empty, aka `interval.low<interval.high`